Add alignment in the Counter of ProfileEvents to reduce false sharing

jiebinn opened the pull request:
This patch will add alignment in the Counter of ProfileEvents to reduce false sharing in ProfileEvents.

Original Clickbench Q3 cycle hotspot (2x240 vCPUs)

![image](https://github.com/user-attachments/assets/a19030e9-f124-4a5b-85ec-4c4ed0595a0b)
![image](https://github.com/user-attachments/assets/39f82e1b-8815-4db0-addd-629f9f2dd022)

Clickbench Q3 cycle hotspot with this patch (2x240 vCPUs)
![image](https://github.com/user-attachments/assets/e428fe1f-ec71-4c83-8b41-dea53e50481d)
The hotspot `ProfileEvents::increment` has reduced from 36.6% to 8.5%.

Clickbench Q3 Core scaling figure (QPS & vCPUs). The patch will improve the performance when the vCPUs larger than 100. Overall, there is 20% ~ 40% performance gain if the vCPUs number is between 100 ~ 480.

![image](https://github.com/user-attachments/assets/28fee3c6-8499-4a15-ad93-d688d74a46fb)


Test the patch with ClickBench Q3 on 2 x 240 vCPUs system. QPS has got 27.4% performance imrpovement and the geometric mean of 43 queries has got 4.4% gain.


<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add alignment in the Counter of ProfileEvents to reduce false sharing.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

shankar-iyer commented:
I am starting the review, please check the test failures.

Quick question.- This technique could be useful for other counters also (current or in future). Can the class and implementation be moved to a higher, common layer maybe inside `ProfileEvents` itself? If inside `ProfileEvents`, the flush can be implicitly done when a profile event is read.

shankar-iyer commented:
Just focusing on -
```
2025-07-02 20:18:42 --- /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/02770_async_buffer_ignore.reference	2025-07-02 20:15:40.190714002 +1100
2025-07-02 20:18:42 +++ /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/02770_async_buffer_ignore.stdout	2025-07-02 20:18:42.898938540 +1100
2025-07-02 20:18:42 @@ -1 +1 @@
2025-07-02 20:18:42 -4	66446	66446
2025-07-02 20:18:42 +4	66446	0
```

Looks like the  `ReadCompressedBytes` counter is not getting updated/flushed **before** the read. I think finding the right time point to flush a batch'ed counter is the challenge. For example, a couple of _static_ flush time points are -

1. For a atomic counter updated repeatedly in a tight loop - just maintain a local variable and update the atomic counter at the end of the loop

2. For Clickhouse, end of the query execution is a good time-point to update a batch'ed counter.

3. When the thread in a thread-pool switches to processing a new query. The thread local destructor may never be entered.

Is there a design pattern where a read of a counter can first update itself from all the thread local counters? Note that in Clickhouse, **real-time** profile events are a key to monitor the progress of a long-running query. You can see that by pressing <space> when a long running query is executing. So that has to be factored in relating to point 2 above. The threshold based flush should be good for this I think.

if it helps, we can even explore declaring batch;ed counters in a different style e.g -

`    M(ReadCompressedBytes, "Number of bytes (the number of bytes before decompression) read from compressed sources (files, network).", ValueType::Bytes) \`

to
  
 ` M_BATCHED(ReadCompressedBytes, "Number of bytes (the number of bytes before decompression) read from compressed sources (files, network).", ValueType::Bytes) \`

>Clickbench Q3 Core scaling figure (QPS & vCPUs). The patch will improve the performance when the vCPUs larger than 100. Overall, there is 20% ~ 40% performance gain if the vCPUs number is between 100 ~ 480.

Fantastic!

jiebinn commented:
> Just focusing on -
> 
> ```
> 2025-07-02 20:18:42 --- /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/02770_async_buffer_ignore.reference	2025-07-02 20:15:40.190714002 +1100
> 2025-07-02 20:18:42 +++ /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/02770_async_buffer_ignore.stdout	2025-07-02 20:18:42.898938540 +1100
> 2025-07-02 20:18:42 @@ -1 +1 @@
> 2025-07-02 20:18:42 -4	66446	66446
> 2025-07-02 20:18:42 +4	66446	0
> ```
> 
> Looks like the `ReadCompressedBytes` counter is not getting updated/flushed **before** the read. I think finding the right time point to flush a batch'ed counter is the challenge. For example, a couple of _static_ flush time points are -
> 
> 1. For a atomic counter updated repeatedly in a tight loop - just maintain a local variable and update the atomic counter at the end of the loop
> 2. For Clickhouse, end of the query execution is a good time-point to update a batch'ed counter.
> 3. When the thread in a thread-pool switches to processing a new query. The thread local destructor may never be entered.
> 
> Is there a design pattern where a read of a counter can first update itself from all the thread local counters? Note that in Clickhouse, **real-time** profile events are a key to monitor the progress of a long-running query. You can see that by pressing when a long running query is executing. So that has to be factored in relating to point 2 above. The threshold based flush should be good for this I think.
> 
> if it helps, we can even explore declaring batch;ed counters in a different style e.g -
> 
> ` M(ReadCompressedBytes, "Number of bytes (the number of bytes before decompression) read from compressed sources (files, network).", ValueType::Bytes) \`
> 
> to
> 
> ` M_BATCHED(ReadCompressedBytes, "Number of bytes (the number of bytes before decompression) read from compressed sources (files, network).", ValueType::Bytes) \`
> 
> > Clickbench Q3 Core scaling figure (QPS & vCPUs). The patch will improve the performance when the vCPUs larger than 100. Overall, there is 20% ~ 40% performance gain if the vCPUs number is between 100 ~ 480.
> 
> Fantastic!

Thanks @shankar-iyer , I will implement it in the original Counters and also use a "pull mode" instead of "push mode".

shankar-iyer commented:
> @shankar-iyer I will let you know once the final solution is ready.

Sure, thanks!

jiebinn committed:
Reduce the contention in atomic of ProfileEvents

This patch will introduce a thread-safe atomic counter that uses sharding to
reduce contention in ProfileEvents. Instead of having all threads compete for
a single atomic variable, this implementation distributes operations across
multiple shards.

Test the patch with ClickBench Q3 on 2 x 240 vCPUs system. QPS has got 27.4%
performance imrpovement and the geometric mean of 43 queries has got 4.4% gain.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>
Reviewed-by: Wangyang Guo <wangyang.guo@intel.com>
Reviewed-by: Tianyou Li <tianyou.li@intel.com>

jiebinn committed:
fit data type

jiebinn committed:
add alignas(64) for Counter as several events will be in the same cacheline and frequently updated

jiebinn committed:
Merge branch 'ClickHouse:master' into profileEvent

jiebinn committed:
Merge branch 'ClickHouse:master' into profileEvent

jiebinn commented:
Hi @shankar-iyer , I have some findings from the investigation and would appreciate your suggestions.

I have completed the initial [patch](https://github.com/ClickHouse/ClickHouse/pull/82697/commits/dfffbb0575a5bdac9409985e2f5320e8894794fa), which replaces `std::atomic<size_t>` with `ShardedAtomicCounter`. This change updates 16 alignas(64) atomic values instead of a single non-alignas atomic value. The `load()` method is used to aggregate all atomic values only when reading.

I experimented with tuning NUM_SHARDS from 16 to 8, 4, 2, and 1. The patch shows a 20% performance gain even when NUM_SHARDS is set to 1. The hotspot `ProfileEvents::increment` has reduced from 36% to 9%. The key difference is the addition of `alignas(64)`. The previous issue was that `ReadCompressedBytes`, `CompressedReadBufferBlocks`, and `CompressedReadBufferBytes` were initialized on the same cacheline, causing significant performance impact due to false sharing.
```
    M(ReadCompressedBytes, "Number of bytes (the number of bytes before decompression) read from compressed sources (files, network).", ValueType::Bytes) \
    M(CompressedReadBufferBlocks, "Number of compressed blocks (the blocks of data that are compressed independent of each other) read from compressed sources (files, network).", ValueType::Number) \
    M(CompressedReadBufferBytes, "Number of uncompressed bytes (the number of bytes after decompression) read from compressed sources (files, network).", ValueType::Bytes) \
```

We have two potential solutions:
1. Make the type Counter alignas(64).
2. Change the location of the three events `ReadCompressedBytes`, `CompressedReadBufferBlocks`, and `CompressedReadBufferBytes` to initialize them in different cachelines.

The patch implements the first solution. There are several micro benchmark [regressions](https://s3.amazonaws.com/clickhouse-test-reports/PRs/82697/8b233c87fd8f846262598c45ba0506fff6a58108//performance_comparison_amd_release_master_head_2_3/report.html), but I wonder if these are caused by variance.

alexey-milovidov committed:
Update ProfileEvents.h

azat commented:
>However, I expect instead of making every counter aligned by 64 bytes, we can create multiple copies of the whole array, std::unique_ptr<Counter[]> counters_holder, to shard by CPU. As these arrays are large, we can avoid false sharing even without extra alignment.

@alexey-milovidov The work started here - https://github.com/ClickHouse/ClickHouse/pull/105056
